distribution/tools/entitlement-agent/src/main/java/org/elasticsearch/entitlement/agent/EntitlementAgent.java

```diff
@@ -9,13 +9,51 @@
 
 package org.elasticsearch.entitlement.agent;
 
-import org.elasticsearch.entitlement.runtime.api.EntitlementChecks;
+import org.elasticsearch.core.SuppressForbidden;
+import org.elasticsearch.entitlement.checks.EntitlementChecks;
+import org.elasticsearch.entitlement.instrumentation.Instrumenter;
 
+import java.io.File;
+import java.io.IOException;
 import java.lang.instrument.Instrumentation;
+import java.util.List;
+import java.util.jar.JarFile;
+
+import static java.util.stream.Collectors.toSet;
 
 public class EntitlementAgent {
 
     public static void premain(String agentArgs, Instrumentation inst) throws Exception {
-        EntitlementChecks.setAgentBooted();
+        // Add the trampoline library (the one with the entitlement checking interface) to the bootstrap classpath.
+        // We can't actually reference the classes here for real before this point because they won't resolve.
+        var jarsString = System.getProperty("entitlements.trampolineJars");
+        if (jarsString != null) {
+            addJarsToBootstrapClassLoader(inst, jarsString);
+        }
+
+        ConfigurationScanner.ScanResults config = ConfigurationScanner.scan(List.of(EntitlementChecks.class));
+
+        inst.addTransformer(
+            new Transformer(
+                new Instrumenter("", config.methodMap()),
+                config.classesToInstrument().stream().map(EntitlementAgent::internalName).collect(toSet())
+            ),
+            true
+        );
+        inst.retransformClasses(config.classesToInstrument().toArray(new Class<?>[0]));
+    }
+
+    @SuppressForbidden(reason = "The appendToBootstrapClassLoaderSearch method takes a JarFile")
+    private static void addJarsToBootstrapClassLoader(Instrumentation inst, String jarsString) throws IOException {
+        for (var jar : jarsString.split(File.pathSeparator)) {
```

Comment:
The first time I tried this, there were two jars due to how I organized the code; plus, Gradle makes it more natural to grab "the files" rather than "the file".

Agreed that we should probably just error out if there are multiple jars.